add plugin content (docs)

Omar-Khawaja opened the pull request:
WIP branch introducing documentation and guide for pluggable drivers, client configuration, and new syntax.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/nomad?pullRequest=5186)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/nomad?pullRequest=5186)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/nomad?pullRequest=5186).</sub>

Omar-Khawaja commented:
Went through built-in drivers docs and updated them with new plugin options + syntax. Called out pluggable drivers in the main task drivers page. Also added **External Plugins** section with updated lxc driver documentation. Added redirect for old lxc driver doc to the new one.

preetapan commented on website/source/docs/external/lxc.html.md:
```diff
@@ -0,0 +1,152 @@
+---
```
Shouldn't this be under docs/drivers/external?

Omar-Khawaja replied:
You're right. Fix has been implemented.

notnoop commented on website/source/docs/external/lxc.html.md:
```diff
@@ -0,0 +1,152 @@
+The `lxc` driver provides an interface for using LXC for running application
+containers. You can download the external LXC driver [here][lxc-driver].
+
+~> Nomad 0.9 does not maintain backward compatibility for the external LXC driver plugin when it comes to client configuration syntax. With Nomad 0.9, you must use the new [plugin syntax][plugin]. See [plugin options][plugin-options] below for an example.
```
I would suggest being more explicit about what changed and what hasn't: e.g. "LXC client setup changed in 0.9: You must use the [plugin syntax][plugin] and install the external lxc driver prior to upgrading.  Job specification remained the same".

notnoop commented on website/source/docs/drivers/lxc.html.md:
```diff
@@ -108,8 +108,10 @@ The `lxc` driver requires the following:
+options](/docs/configuration/client.html#options):
+
 * `lxc.enable` - The `lxc` driver may be disabled on hosts by setting this
-  [client configuration][/docs/configuration/client.html##options-parameters]
```
This file is not relevant anymore and probably should be deleted?!

schmichael commented on website/source/docs/configuration/client.html.md:
```diff
@@ -28,6 +28,10 @@ client {
 
+## Plugin Options
+
+Nomad 0.9 now supports pluggable drivers. Operators should use the new [plugin][plugin-stanza] syntax to modify driver configuration. To find the plugin options supported by each individual Nomad driver, please see the [drivers documentation](/docs/drivers/index.html).
```
Maybe add a line like this since we expect people to upgrade to 0.9 slowly (and therefore have lots of docs visitors who care about pre-0.9 things):
> The pre-0.9 `client.options` stanza will be supported in 0.9 for backward compatibility but will be removed in a future release.

schmichael commented on website/source/docs/configuration/client.html.md:
```diff
@@ -158,6 +162,8 @@ the full list.
 
 ### `options` Parameters
 
+~> Note: client configuration options for drivers will soon be deprecated.. Please use [plugin options][plugin-options] instead. See the [plugin stanza][plugin-stanza] documentation for more information.
```
It seems a bit strange to send them to `plugin options` which just sends them to `plugin stanza` as well? Maybe just remove the middle sentence or maybe I'm missing something.

```suggestion
~> Note: client configuration options for drivers will soon be deprecated. See the [plugin stanza][plugin-stanza] documentation for more information.
```

schmichael commented on website/source/docs/configuration/plugin.html.md:
```diff
@@ -17,7 +17,7 @@ description: |-
 </table>
 
-The `plugin` stanza is used to configure an individual plugin.
+The `plugin` stanza is used to configure an individual plugin. You may use this section to introduce new task or device drivers or simply modify their options.
```
I don't think you can use the `plugin` stanza to introduce/install/enable plugins. I think they're either auto-detected in the plugin directory or not?

```suggestion
The `plugin` stanza is used to configure plugins.
```

schmichael commented on website/source/docs/drivers/docker.html.md:
```diff
@@ -479,10 +480,12 @@ Example agent configuration, using a helper script "docker-credential-ecr" in
-  enabled = true
-  options {
-    "docker.auth.helper" = "ecr"
+client {...}
```
Let's leave a functional client stanza in place to support copying and pasting.

```suggestion
client {
  enabled = true
}

```

schmichael commented on website/source/docs/drivers/docker.html.md:
```diff
@@ -614,8 +617,105 @@ user to the `docker` group so you can run Nomad without root:
+    allow_privileged = false
+    allow_caps       = ["CHOWN", "NET_RAW"]
+
+    # values can also be set to "ALL"
```
```suggestion
    # allow_caps can also be set to "ALL"
```

schmichael commented on website/source/docs/drivers/docker.html.md:
```diff
@@ -614,8 +617,105 @@ user to the `docker` group so you can run Nomad without root:
+
 ## Client Configuration
 
+~> Note: client configuration options will soon be deprecated. Please use [plugin options][plugin-options] instead. See the [plugin stanza][plugin-stanza] documentation for more information.
```
Another spot maybe we can just shorten it to `plugin stanza` as Plugin Options has very little content?

Omar-Khawaja replied:
Thanks for this tip, Michael. I implemented this suggestion in the main client configuration doc, but the plugin options links in the individual task drivers actually link back to more information/options, so I think we should keep it for at least those docs? If you take another pass at this section in particular and still feel we need to shorten, I'll be happy to make the change.

schmichael replied:
Sounds good to me! :+1:

schmichael commented on website/source/docs/drivers/docker.html.md:
```diff
@@ -677,8 +777,8 @@ options](/docs/configuration/client.html#options):
-  `"CHOWN,DAC_OVERRIDE,FSETID,FOWNER,MKNOD,NET_RAW,SETGID,SETUID,SETFCAP,SETPCAP,NET_BIND_SERVICE,SYS_CHROOT,KILL,AUDIT_WRITE"`,
-  which is the list of capabilities allowed by docker by default, as
+  `"CHOWN,DAC_OVERRIDE,FSETID,FOWNER,MKNOD,NET_RAW,SETGID,SETUID,SETFCAP,
+   SETPCAP,NET_BIND_SERVICE,SYS_CHROOT,KILL,AUDIT_WRITE"`, which is the list of  capabilities allowed by docker by default, as
```
```suggestion
   SETPCAP,NET_BIND_SERVICE,SYS_CHROOT,KILL,AUDIT_WRITE"`, which is the list of capabilities allowed by docker by default, as
```

schmichael commented on website/source/docs/drivers/index.html.md:
```diff
@@ -12,6 +12,8 @@ Task drivers are used by Nomad clients to execute a task and provide resource
 isolation. By having extensible task drivers, Nomad has the flexibility to
 support a broad set of workloads across all major operating systems.
 
+Starting with Nomad 0.9, task drivers are now pluggable. This gives users the flexibility to introduce their own drivers without having to recompile Nomad. You can view the [plugin stanza][plugin] documentation for examples on how to use the `plugin` stanza in Nomad's client configuration. Note that we have introduced new syntax when specifying driver options in the client configuration (see [doc ...
```
...except lxc :(

As much as I want to ignore it we still need to mention it here I think.

schmichael commented on website/source/docs/drivers/raw_exec.html.md:
```diff
@@ -89,10 +99,24 @@ client {
+
 ## Client Options
 
+~> Note: client configuration options will soon be deprecated.. Please use 
```
Copy and paste from other areas if you end up making edits.

Omar-Khawaja commented on website/source/docs/drivers/external/lxc.html.md:
```diff
@@ -0,0 +1,147 @@
+isolation is not supported as of now.
+
+[lxc-create]: https://linuxcontainers.org/lxc/manpages/man1/lxc-create.1.html
+[lxc-driver]: /placeholder/for/lxc-driver/link
```
Thank you. Just updated the official link.
